packages/sanity/src/core/store/_legacy/document/createPairListener.ts

```diff
@@ -141,6 +150,99 @@ export function getPairListener(
+  client: SanityClient,
+  idPair: IdPair,
+  options?: PairListenerOptions,
+): Observable<ClientListenerEvent> {
+  const {publishedId, draftId} = idPair
+  const currentLeg = defer(
+    () =>
+      client.observable.listen(
+        `*[_id == $publishedId || _id == $draftId]`,
+        {publishedId, draftId},
+        {
+          includeResult: false,
+          events: ['welcome', 'mutation', 'reconnect'],
+          effectFormat: 'mendoza',
+          tag: options?.tag || 'document.pair-listener',
+        },
+      ) as Observable<ClientListenerEvent>,
+  ).pipe(share())
+
+  if (!options?.relay) {
+    return currentLeg
+  }
+
+  const {exchangeWaitMin, exchangeWaitMax, exchangeOverLapTime, exchangeTimeout} = options.relay
+
+  // This represents the next leg, and will be started after a certain delay
+  const nextLeg = currentLeg.pipe(
+    // make sure we are connected to the current leg before scheduling the next one. This prevents build-up in case it takes a while to connect
+    filter((ev) => ev.type === 'welcome'),
+    // current listener may still get disconnected and reconnect,
+    // so in case we receive a new welcome event we should cancel the previously scheduled next leg
+    switchMap(() => timer(getExchangeWait(exchangeWaitMin, exchangeWaitMax))),
+    take(1),
+    mergeMap(() =>
+      createRelayPairListener(client, idPair, options).pipe(
+        // this will make sure that if it takes too long to connect to the next, we will just ignore it and continue with the current leg
+        timeout({
+          first: exchangeTimeout,
+          with: () => EMPTY,
+        }),
+      ),
+    ),
+    share(),
+  )
+
+  // Merge messages from current leg with next leg into a single stream
+  return merge(
+    // take from currentLeg until we get the first 'welcome' from next leg, plus adding little overlap in time
+    currentLeg.pipe(
+      takeUntil(
+        nextLeg.pipe(
+          filter((e) => e.type === 'welcome'),
+          take(1),
+          delay(exchangeOverLapTime),
+        ),
+      ),
+    ),
+    // ignore the first welcome event from the next leg.
+    nextLeg.pipe(filter((e, index) => e.type !== 'welcome' || index > 0)),
+  ).pipe(distinctByEventId())
+}
+
+/**
+ * Operator function that takes a stream of listener events
+ * and returns a new stream that filters out events sharing the same eventId
+ */
+function distinctByEventId() {
+  return (input$: Observable<ClientListenerEvent>) =>
+    input$.pipe(
+      scan(
+        (
+          [seen]: [LRU<string, boolean>, ClientListenerEvent | null],
+          event,
+        ): [LRU<string, boolean>, ClientListenerEvent | null] => {
+          if (event.type !== 'mutation') {
+            return [seen, event]
+          }
+          if (seen.has(event.eventId)) {
+            return [seen, null]
+          }
+          seen.set(event.eventId, true)
+          return [seen, event]
+        },
+        [new LRU<string, boolean>({maxSize: 1000}), null],
```

Comment:
To make sure I understand: this means we'll buffer up to 1,000 messages per listener to deduplicate during the 20 second exchange overlap. Is my understanding correct?

Related:

1. I assume the memory impact of buffering (up to) 1,000 messages per listener is inconsequential. If this number was picked arbitrarily, could it be beneficial to decrease it?
2. I assume that it's possible (but unlikely) for duplicate events to be emitted if more than 1,000 messages are received during the 20 second exchange overlap. Is it problematic if that occurs? My guess is everything would be okay, and the purpose of deduping is more to ensure listeners behave in the way consumers expect.

Reply:
Great questions.

> Is my understanding correct?

Yes. The key here is "up to" - in reality it's unlikely to be more than a handful events.

> I assume the memory impact of buffering (up to) 1,000 messages per listener is inconsequential. If this number was picked arbitrarily, could it be beneficial to decrease it?

I'd expect that to be extremely rare, so 1000 felt more like a "safe upper limit"

> I assume that it's possible (but unlikely) for duplicate events to be emitted if more than 1,000 messages are received during the 20 second exchange overlap. Is it problematic if that occurs?

If it occurs it will likely lead to problems (the same diffMatchPatch applied twice leads to wrong content), so you made me realize this buffer should be time bound not size bound. I'll refactor this. Thanks!

Reply:
Got it. Thanks for explaining!

Reply:
Pushed a refactor in a8dc8b1c6ba203e697fc0c3839f7fa03f5e3a0c1 that will dedupe listener events based on their eventId for the whole of the overlap period duration. When the overlap period ends, it will clear the set of seen event ids and stop the deduping.